instrumentation/undertow-1.4/src/main/java/io/opentelemetry/javaagent/instrumentation/hypertrace/undertow/v1_4/utils/Utils.java

```diff
@@ -0,0 +1,97 @@
+   *
+   * @param httpServerExchange that {@link HttpServerExchange#getRequestChannel()} was invoked on
+   * @param returnedChannel {@link StreamSourceChannel} implementation created and held by {@link
+   *     HttpServerExchange} for the request body
+   * @param contextStore to hold the {@link SpanAndBuffer} associated with the {@link
+   *     StreamSourceChannel} returned by {@link HttpServerExchange#getRequestChannel()}
+   */
+  public static void createAndStoreBufferForSpan(
+      final HttpServerExchange httpServerExchange,
+      final StreamSourceChannel returnedChannel,
+      final ContextStore<StreamSourceChannel, SpanAndBuffer> contextStore) {
+    if (!ConfigProvider.get().httpBody().request()
+        || !ContentTypeUtils.shouldCapture(
+            httpServerExchange.getRequestHeaders().getFirst(Headers.CONTENT_TYPE))) {
+      return;
+    }
+    final Span span =
+        Span.fromContext(UndertowHttpServerTracer.tracer().getServerContext(httpServerExchange));
+    final Charset charset = Charset.forName(httpServerExchange.getRequestCharset());
+    final BoundedByteArrayOutputStream boundedByteArrayOutputStream =
+        BoundedBuffersFactory.createStream(
+            (int) httpServerExchange.getRequestContentLength(), charset);
+    final SpanAndBuffer spanAndBuffer =
+        new SpanAndBuffer(
+            span,
+            boundedByteArrayOutputStream,
+            HypertraceSemanticAttributes.HTTP_REQUEST_BODY,
+            charset);
+    contextStore.put(returnedChannel, spanAndBuffer);
+    httpServerExchange.addExchangeCompleteListener(
+        new BodyCapturingExchangeCompletionListener(spanAndBuffer));
+  }
+
+  /**
+   * @param readOnlyBuffer that was just read into {@link StreamSourceChannel#read(ByteBuffer)}
+   * @param numBytesRead into the provided {@link ByteBuffer} by {@link
+   *     StreamSourceChannel#read(ByteBuffer)}
+   * @param spanAndBuffer named tuple retrieved from the {@link
+   *     io.opentelemetry.javaagent.instrumentation.api.InstrumentationContext} where the {@link
+   *     StreamSourceChannel} instance is the key. This was put into the {@link
+   *     io.opentelemetry.javaagent.instrumentation.api.InstrumentationContext} by {@link
+   *     io.opentelemetry.javaagent.instrumentation.hypertrace.undertow.v1_4.UndertowHttpServerExchangeInstrumentation}
+   *     the first time {@link HttpServerExchange#getRequestChannel()} was invoked on the request
+   */
+  public static void handleRead(
+      final ByteBuffer readOnlyBuffer, final int numBytesRead, final SpanAndBuffer spanAndBuffer) {
+    if (numBytesRead <= 0) {
+      return;
+    }
+    readOnlyBuffer.position(readOnlyBuffer.position() - numBytesRead);
+
+    final BoundedByteArrayOutputStream boundedByteArrayOutputStream = spanAndBuffer.byteArrayBuffer;
+    for (int i = 0; i < numBytesRead; i++) {
+      final byte b = readOnlyBuffer.get();
+      boundedByteArrayOutputStream.write(b);
```

Comment:
This does not seem efficient, `ByteBuffer` should have an API to return the allocated array.

Reply:
`ByteBuffer.array()` does exist, but it's not guaranteed to work (depending on the kind of `ByteBuffer` being used. As far as I know, it does not work on `DirectByteBuffer`s, only on heap-based buffers. In the case of undertow, the `ByteBuffer` subclass being used in the test case is a `DirectByteBuffer` so `ByteBuffer.array()` will throw an `UnsupportedOperationException`. 

My first pass at this, after discovering that `ByteBuffer.array()` does not suit our needs, was to do something like:
```java
byte[] dst = new byte[numBytesRead];
readOnlyBuffer.get(dst);
boundedByteArrayOutputStream.write(dst);
```

However, after doing this, I looked at the implementation of `ByteBuffer.get(byte[], int, int)`. It looks like this:

```    
public ByteBuffer get(byte[] dst, int offset, int length) {
  checkBounds(offset, length, dst.length);
  if (length > remaining())
    throw new BufferUnderflowException();
  int end = offset + length;
  for (int i = offset; i < end; i++)
      dst[i] = get();
      return this;
}
```

Note `ByteBuffer.get()` is called `n` times, as it is in this method as well. So, by calling `ByteBuffer.get()` directly here, it is invoked the same number of times as `ByteBuffer.get(byte[])`, except that we get to avoid allocating a duplicate `byte[]` just to have it eligible for GC at the end of this utility method. This could put additional GC pressure on the application and result in degraded performance. I'm definitely open to other options of copying data from the `ByteBuffer`! But as far as I know, this is our best option today.